lib/pry.rb

```diff
@@ -46,6 +46,15 @@ class Pry
     else
       output.puts "#{exception.class}: #{exception.message}"
       output.puts "from #{exception.backtrace.first}"
+
+      if exception.respond_to? :cause
+        cause = exception.cause
+        while cause
```

Comment:
I believe it would be better to limit this condition, say, to 3 nested exceptions. It will guard against the case when an exception got stuck in a weird loop and generated a lot of causes. With the check you wrote we would probably be stuck in the loop for a long time. With the additional check we would exit soon.

Reply:
Same idea applies to the other file. Because of this additional logic, I would probably create a new simple class that wraps this logic, so we don't repeat ourselves.

Reply:
Setting a limit will not be helpful as far as I can tell.

- Empirical evidence given by Java is that limit has never been needed.
- Usually inner-most exception is most interesting so eating any layers would be undesirable.
- Listing nested exceptions is not on the performance critical path. It only happens when execution should stop anyway. And it is **fast**. I tried 1000 nested exceptions and it appears to work instantaneously for me:

```
[21] pry(main)> 1000.times { |i|
[21] pry(main)*   ex = begin
[21] pry(main)*     raise ex
[21] pry(main)*   rescue  
[21] pry(main)*     raise "parent#{i}"
[21] pry(main)*   end rescue $!  
[21] pry(main)* }  
=> 1000
[22] pry(main)> ex.cause.cause
=> #<RuntimeError: parent97>
...
[37] pry(main)> wtf?!?!?!
...
Caused by: RuntimeError: parent2
--
 0: (pry):22:in `rescue in block in __pry__'
 1: (pry):19:in `block in __pry__'
 2: (pry):18:in `times'
 3: (pry):18:in `__pry__'
 4: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_instance.rb:355:in `eval'
 5: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_instance.rb:355:in `evaluate_ruby'
 6: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_instance.rb:323:in `handle_line'
 7: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_instance.rb:243:in `block (2 levels) in eval'
 8: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_instance.rb:242:in `catch'
 9: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_instance.rb:242:in `block in eval'
10: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_instance.rb:241:in `catch'
11: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_instance.rb:241:in `eval'
12: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/repl.rb:77:in `block in repl'
13: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/repl.rb:67:in `loop'
14: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/repl.rb:67:in `repl'
15: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/repl.rb:38:in `block in start'
16: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/input_lock.rb:61:in `__with_ownership'
17: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/input_lock.rb:79:in `with_ownership'
18: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/repl.rb:38:in `start'
19: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/repl.rb:13:in `start'
20: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/pry_class.rb:192:in `start'
21: /home/avalon/.gem/ruby/gems/pry-0.11.3/lib/pry/cli.rb:116:in `start'
22: /home/avalon/.gem/ruby/gems/pry-0.11.3/bin/pry:12:in `<top (required)>'
23: /home/avalon/bin/pry:23:in `load'
24: /home/avalon/bin/pry:23:in `<main>'
Caused by: RuntimeError: parent1
--
 0: (pry):22:in `rescue in block in __pry__'
...
```

Reply:
It's hard for me to reason about Java because I am not sure what you mean and I've never used Java for anything serious myself :)

I feel like if you nest exceptions purposefully and you have more than 3 causes, you are already screwed. My wording was probably a little bit wrong. My point is not the performance but the usefulness of viewing a long list of causes, which is practically impossible to grasp.

Let's imagine one `SystemStackError`:

```ruby
def foo; foo; end
foo
```

This raises around 13101 frames for me. It doesn't have a nested exception but let's imagine it does. That would be 2*13101. 98% of the frames will be repetitive.

I believe a limit of 3-5 causes is enough and in the future we could relax that. I also believe this is an extreme edge case, so the limit is more like a safety measure.

Reply:
Nesting exceptions depends on layers of abstraction, not recursion. e.g. your program reads configuration in YAML format. And naturally uses a library for reading the YAML. Then the library raises some error `file not found`. Then you raise error `error reading configuration` with cause whatever the underlying library returned. Thus user sees:
```
error reading configuration <- file not  found
```

In our code we handle any error with the same code, just taking whatever cause underlying library gave us. If it raised `syntax error`, then user would have seen:
```
error reading configuration <- syntax error
```

Then one can inspect the error at the different abstraction layers.

What I gave as an example of 1000 nested exceptions was deliberately implemented for performance check. What I said about Java is that there nested exceptions are a heavily used feature maybe since the language has been invented. And I never heard or anybody wanting to limit reporting of nesting. Although This is some real-world example 

Additionally, if I see 100 nested errors I will know something is terribly wrong with error handling code somewhere. So I would look into the repeated errors to figure out what's wrong. If I only see 3 errors, then I may think that this is all and keep broken error handling code around.

I hope this clarifies why am I against setting a limit and don't feel motivated to implement such. For me personally having a limit of 1 or 3 is more or less the same bug.

If we really want to short output, then we can keep existing logic to only show one backtrace, but make sure to print out a line (possibly in another color, that error has nested causes. Then user can request additional causes somehow, but if I request causes I'd want all of them, not some limit.

Reply:
Fair enough, sounds good. I am not feeling strongly about this because it won't likely be a common thing to stumble upon. For this reason I think the limit can be omitted. If anything, it'd be trivial to introduce it.